File: solutions/300/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Nov 26 10:51:33 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <algorithm> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // dp table
        // initialize
        vector<int> dp(nums.size(), 1);
        // traverse dp table
        for (int i = 1; i < nums.size(); ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] > nums[j])
                    dp[i] = max(dp[i], dp[j] + 1);
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};

TEST(Test300, SimpleTest) {
    vector<int> nums{
        10,9,2,5,3,7,101,18
    };
    Solution s; 
    EXPECT_EQ(s.lengthOfLIS(nums), 4); 
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/300/main.py
from typing import List 

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # check input
        if len(nums) == 1:
            return 1
        # dp table
        # initialize
        dp = [1] * len(nums)
        # traverse dp table
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(*dp)

if __name__ == "__main__": 
    nums = [10,9,2,5,3,7,101,18]
    s = Solution()
    assert s.lengthOfLIS(nums) == 4
